src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj

```diff
@@ -83,43 +79,57 @@
-            (finally
-              (.releaseItem (:pool old-pool) instance false)))
-          (log/infof "Finished creating JRubyPuppet instance %d of %d"
-                     id count))
-        (catch Exception e
-          (.clear new-pool)
-          (.insertPill new-pool (PoisonPill. e))
-          (throw (IllegalStateException.
-                   "There was a problem adding a JRubyPuppet instance to the pool."
-                   e)))))
-    ;; Add a "RetryPoisonPill" to the pool in case something else is in the
-    ;; process of borrowing from the old pool.
-    (.insertPill (:pool old-pool) (RetryPoisonPill. (:pool old-pool)))))
+  (try
+    (log/info "Flush request received; creating new JRuby pool.")
+    (let [{:keys [config profiler pool-state]} pool-context
+          new-pool-state (jruby-internal/create-pool-from-config config)
+          new-pool (:pool new-pool-state)
+          old-pool @pool-state
+          old-pool-size (:size old-pool)]
+      (log/info "Replacing old JRuby pool with new instance.")
+      (if (and (= reason :shutdown) (not= old-pool-size (count (.getRegisteredElements (:pool old-pool)))))
+        (log/warn "Attempting to flush a pool that does not appear to have successfully initialized. Aborting.")
+        (do
+          (when (= reason :shutdown)
+            (.insertPill new-pool (ShutdownPoisonPill. new-pool)))
+          (reset! pool-state new-pool-state)
+          (log/info "Swapped JRuby pools, beginning cleanup of old pool.")
+          (doseq [i (range old-pool-size)]
+            (try
+              (let [id (inc i)
+                    instance (jruby-internal/borrow-from-pool!*
+                              jruby-internal/borrow-without-timeout-fn
+                              (:pool old-pool))]
+                (try
+                  (jruby-internal/cleanup-pool-instance! instance)
+                  (when (= reason :flush-requested)
+                    (jruby-internal/create-pool-instance! new-pool id config
+                                                          (partial send-flush-instance! pool-context)
+                                                          profiler)
+                    (log/infof "Finished creating JRubyPuppet instance %d of %d"
+                               id old-pool-size))
+                  (finally
+                    (.releaseItem (:pool old-pool) instance false))))
+              (catch Exception e
+                (.clear new-pool)
+                (.insertPill new-pool (PoisonPill. e))
+                (throw (IllegalStateException.
+                        "There was a problem adding a JRubyPuppet instance to the pool."
+                        e)))))
+          ;; Add a "RetryPoisonPill" to the pool in case something else is in the
+          ;; process of borrowing from the old pool.
+          (.insertPill (:pool old-pool) (RetryPoisonPill. (:pool old-pool))))))
+    (finally
+      (deliver flush-complete? true))))
```

Comment:
There's now a non-trivial amount of code in this function which is only applicable to one of two cases: flushing the pool for a shutdown, or flushing and re-creating the pool while the server is running.  Instead of simply extending this function to handle both cases, I wonder if it would be cleaner to introduce a new function like `flush-pool-for-shutdown!`  (and/or perhaps name this to `flush-and-repopulate-pool!`) and factoring-out bits of this function as-needed for reuse.